src/main/java/com/google/devtools/build/lib/bazel/repository/downloader/HttpDownloader.java

```diff
@@ -73,6 +77,17 @@ public void setMaxRetryTimeout(Duration maxRetryTimeout) {
     this.maxRetryTimeout = maxRetryTimeout;
   }
 
+  public void setMaxParallelDownloads(int maxParallelDownloads) {
```

Comment:
Is there a way to refactor this further so that we can pass this value into the `HttpDownloader` constructor instead?

Reply:
Not without significant rearchitecting, or another layer of indirection around `HttpDownloader`, or a startup option:
* the remote module is initialized before the repository module, but depends on the httpDownloader and the downloader delegate to be set up by the repository module. The downloader delegate in turn requires an instance of httpDownloader (to set the default delegate). This means that we don't have the `RepositoryOptions` available to set up the downloader until after we need to provide those objects to the remote module (workaround: Use a startup option)
* we're passing the download manager to `VendorCommand` as part of `serverInit`, which again, only gets startup options
* We could introduce a supplier indirection around HttpDownloader, and hope/pray/assert it's not actually used before we set it up in `beforeCommand` - that's pretty opaque though

This would make the semaphore handling code much easier though, so there's no funny release/acquire business needed, so I see why you're asking.

Reply:
I'll attempt moving the entire thing into `BlazeRuntime` and `CommonCommandOptions`, but I suggest we not block this PR on that, given the deadline of 7.3.0.

Reply:
See https://github.com/bazelbuild/bazel/pull/23106/commits/bfc545adef37a3aac114e493ec96840b4754e77f#r1691336135 for an alternative idea that wouldn't require as much refactoring.

Reply:
I did the major refactor now, moving the http downloader+delegate into `CommandEnvironment`, powered by `CommonCommandOptions`. That feels pretty reasonable - http downloading is (apparently) shared by multiple commands and multiple modules, so now those modules don't need to know about each other.
It has the downside of delaying the initialization of `DownloadManager` past the initialization of some objects that depend on it, but that seems to be a recurring pattern in the codebase.